File: app/src/main/java/com/bajaj/rideconnect/re/WavySeekBar.java
package com.bajaj.rideconnect.re;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.drawable.ColorDrawable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.animation.LinearInterpolator;
import android.widget.SeekBar;

/**
 * Android 14/15 Material You inspired Wavy/Squiggly Media Scrubber SeekBar.
 * Features:
 * - Sleek, thin stroke thickness (2.5dp) for high-end automotive styling.
 * - Smooth sine wave ripple animation when music is playing.
 * - Graceful transition to flat straight line when paused.
 * - Smooth envelope damping at wave boundaries (starts and ends seamlessly on centerline).
 * - Full-height tactile touch hit-box (40dp) for effortless tapping and scrubbing.
 * - High-contrast glowing cyan & white thumb handle.
 */
public class WavySeekBar extends SeekBar {

    private final Paint wavePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint bgTrackPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint thumbPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint thumbGlowPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Path wavePath = new Path();

    private float density = 1.0f;
    private float strokeWidth;
    private float waveAmplitude;
    private float waveLength;
    private float currentAmplitude = 0f;
    private float phase = 0f;

    private float thumbRadius;
    private float thumbGlowRadius;

    private boolean isPlaying = false;
    private boolean isDragging = false;

    private ValueAnimator waveAnimator;
    private ValueAnimator amplitudeAnimator;
    private OnSeekBarChangeListener listener;

    public WavySeekBar(Context context) {
        super(context);
        init(context);
    }

    public WavySeekBar(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public WavySeekBar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        density = context.getResources().getDisplayMetrics().density;

        strokeWidth = 2.5f * density;
        waveAmplitude = 3.5f * density;
        waveLength = 22.0f * density;
        thumbRadius = 5.5f * density;
        thumbGlowRadius = 8.5f * density;

        // Wave Paint (Played progress) - Vibrant Neon Cyan
        wavePaint.setColor(Color.parseColor("#38BDF8"));
        wavePaint.setStyle(Paint.Style.STROKE);
        wavePaint.setStrokeWidth(strokeWidth);
        wavePaint.setStrokeCap(Paint.Cap.ROUND);
        wavePaint.setStrokeJoin(Paint.Join.ROUND);

        // Background Track (Unplayed) - Sleek Dark Slate
        bgTrackPaint.setColor(Color.parseColor("#222838"));
        bgTrackPaint.setStyle(Paint.Style.STROKE);
        bgTrackPaint.setStrokeWidth(strokeWidth);
        bgTrackPaint.setStrokeCap(Paint.Cap.ROUND);

        // Thumb Paints
        thumbPaint.setColor(Color.WHITE);
        thumbPaint.setStyle(Paint.Style.FILL);

        thumbGlowPaint.setColor(Color.parseColor("#06B6D4"));
        thumbGlowPaint.setStyle(Paint.Style.FILL);

        // Clear default Android SeekBar drawables to ensure 100% custom canvas rendering
        setProgressDrawable(new ColorDrawable(Color.TRANSPARENT));
        setThumb(new ColorDrawable(Color.TRANSPARENT));
        setSplitTrack(false);
        setBackgroundColor(Color.TRANSPARENT);

        // Default to flat line until playing state is enabled
        currentAmplitude = 0f;
    }

    /**
     * Dynamically updates the Material You accent color for the wave ripple and thumb glow.
     */
    public void setAccentColor(int accentColor) {
        wavePaint.setColor(accentColor);
        thumbGlowPaint.setColor(accentColor);
        invalidate();
    }

    @Override
    public void setOnSeekBarChangeListener(OnSeekBarChangeListener l) {
        super.setOnSeekBarChangeListener(l);
        this.listener = l;
    }

    /**
     * Updates playback state. Starts wavy ripple animation when playing,
     * or smoothly flattens out into a straight line when paused.
     */
    public void setPlaying(boolean playing) {
        if (this.isPlaying == playing) return;
        this.isPlaying = playing;
        animateStateTransition();
    }

    public boolean isPlaying() {
        return isPlaying;
    }

    private void animateStateTransition() {
        if (amplitudeAnimator != null) {
            amplitudeAnimator.cancel();
        }

        float targetAmp = isPlaying ? waveAmplitude : 0f;
        amplitudeAnimator = ValueAnimator.ofFloat(currentAmplitude, targetAmp);
        amplitudeAnimator.setDuration(350);
        amplitudeAnimator.addUpdateListener(animation -> {
            currentAmplitude = (float) animation.getAnimatedValue();
            invalidate();
        });
        amplitudeAnimator.start();

        if (isPlaying) {
            if (waveAnimator == null) {
                waveAnimator = ValueAnimator.ofFloat(0f, (float) (2 * Math.PI));
                waveAnimator.setDuration(1200);
                waveAnimator.setInterpolator(new LinearInterpolator());
                waveAnimator.setRepeatCount(ValueAnimator.INFINITE);
                waveAnimator.addUpdateListener(animation -> {
                    phase = (float) animation.getAnimatedValue();
                    invalidate();
                });
            }
            if (!waveAnimator.isRunning()) {
                waveAnimator.start();
            }
        } else {
            if (waveAnimator != null && currentAmplitude <= 0.05f) {
                waveAnimator.cancel();
            }
        }
    }

    @Override
    protected synchronized void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int desiredHeight = Math.round(36f * density);
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);

        int measuredHeight;
        if (heightMode == MeasureSpec.EXACTLY) {
            measuredHeight = heightSize;
        } else if (heightMode == MeasureSpec.AT_MOST) {
            measuredHeight = Math.min(desiredHeight, heightSize);
        } else {
            measuredHeight = desiredHeight;
        }

        setMeasuredDimension(MeasureSpec.getSize(widthMeasureSpec), measuredHeight);
    }

    @Override
    protected synchronized void onDraw(Canvas canvas) {
        int w = getWidth();
        int h = getHeight();
        if (w <= 0 || h <= 0) return;

        float centerY = h / 2f;
        float trackStart = getPaddingLeft() + thumbGlowRadius;
        float trackEnd = w - getPaddingRight() - thumbGlowRadius;
        float trackWidth = Math.max(1f, trackEnd - trackStart);

        int max = getMax();
        float progressFraction = (max > 0) ? Math.max(0f, Math.min(1f, (float) getProgress() / max)) : 0f;
        float thumbX = trackStart + trackWidth * progressFraction;

        // 1. Draw Unplayed Track (Straight line from thumbX to trackEnd)
        if (thumbX < trackEnd) {
            canvas.drawLine(thumbX, centerY, trackEnd, centerY, bgTrackPaint);
        }

        // 2. Draw Played Track (Wavy Sine Wave from trackStart to thumbX)
        if (thumbX > trackStart) {
            float playedWidth = thumbX - trackStart;
            if (currentAmplitude > 0.1f && playedWidth > 4f * density) {
                wavePath.reset();
                wavePath.moveTo(trackStart, centerY);

                float step = 2.0f * density;
                float rampDist = Math.min(waveLength * 0.75f, playedWidth / 2f);

                for (float x = trackStart; x <= thumbX; x += step) {
                    float distStart = x - trackStart;
                    float distEnd = thumbX - x;
                    float env = 1f;

                    if (rampDist > 0.1f) {
                        float startRamp = Math.min(1f, distStart / rampDist);
                        float endRamp = Math.min(1f, distEnd / rampDist);
                        float minRamp = Math.min(startRamp, endRamp);
                        // Smooth sinusoidal easing envelope
                        env = (float) Math.sin(minRamp * Math.PI / 2.0);
                    }

                    float angle = (float) (2.0 * Math.PI * (x - trackStart) / waveLength - phase);
                    float y = centerY + currentAmplitude * env * (float) Math.sin(angle);
                    wavePath.lineTo(x, y);
                }

                // Ensure perfect connection to thumb point
                wavePath.lineTo(thumbX, centerY);
                canvas.drawPath(wavePath, wavePaint);
            } else {
                // Flat line fallback when paused or very short progress
                canvas.drawLine(trackStart, centerY, thumbX, centerY, wavePaint);
            }
        }

        // 3. Draw Thumb Handle
        float currentGlow = isDragging ? (thumbGlowRadius + 3.0f * density) : thumbGlowRadius;
        int glowAlpha = isDragging ? 180 : 120;
        thumbGlowPaint.setAlpha(glowAlpha);
        canvas.drawCircle(thumbX, centerY, currentGlow, thumbGlowPaint);

        // Inner solid white circle
        canvas.drawCircle(thumbX, centerY, thumbRadius, thumbPaint);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (!isEnabled()) return false;

        float trackStart = getPaddingLeft() + thumbGlowRadius;
        float trackEnd = getWidth() - getPaddingRight() - thumbGlowRadius;
        float trackWidth = Math.max(1f, trackEnd - trackStart);

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                isDragging = true;
                if (getParent() != null) {
                    getParent().requestDisallowInterceptTouchEvent(true);
                }
                updateTouchProgress(event.getX(), trackStart, trackWidth);
                if (listener != null) {
                    listener.onStartTrackingTouch(this);
                }
                invalidate();
                return true;

            case MotionEvent.ACTION_MOVE:
                if (isDragging) {
                    updateTouchProgress(event.getX(), trackStart, trackWidth);
                    invalidate();
                    return true;
                }
                break;

            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                if (isDragging) {
                    updateTouchProgress(event.getX(), trackStart, trackWidth);
                    isDragging = false;
                    if (listener != null) {
                        listener.onStopTrackingTouch(this);
                    }
                    invalidate();
                    return true;
                }
                break;
        }

        return super.onTouchEvent(event);
    }

    private void updateTouchProgress(float x, float trackStart, float trackWidth) {
        float clampedX = Math.max(trackStart, Math.min(trackStart + trackWidth, x));
        float fraction = (clampedX - trackStart) / trackWidth;
        int newProgress = Math.round(fraction * getMax());
        setProgress(newProgress);
        if (listener != null) {
            listener.onProgressChanged(this, newProgress, true);
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (waveAnimator != null) {
            waveAnimator.cancel();
        }
        if (amplitudeAnimator != null) {
            amplitudeAnimator.cancel();
        }
    }
}

File: app/src/com/bajaj/rideconnect/re/WavySeekBar.kt
package com.bajaj.rideconnect.re

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.drawable.ColorDrawable
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.animation.LinearInterpolator
import android.widget.SeekBar
import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin

class WavySeekBar @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : SeekBar(context, attrs, defStyleAttr) {

    private val wavePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val bgTrackPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val thumbPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val thumbGlowPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val wavePath = Path()

    private var density = 1.0f
    private var strokeWidth = 0f
    private var waveAmplitude = 0f
    private var waveLength = 0f
    private var currentAmplitude = 0f
    private var phase = 0f

    private var thumbRadius = 0f
    private var thumbGlowRadius = 0f

    var isPlaying: Boolean = false
        private set

    private var isDragging = false

    private var waveAnimator: ValueAnimator? = null
    private var amplitudeAnimator: ValueAnimator? = null
    private var listener: OnSeekBarChangeListener? = null

    init {
        density = context.resources.displayMetrics.density

        strokeWidth = 2.5f * density
        waveAmplitude = 3.5f * density
        waveLength = 22.0f * density
        thumbRadius = 5.5f * density
        thumbGlowRadius = 8.5f * density

        wavePaint.color = Color.parseColor("#38BDF8")
        wavePaint.style = Paint.Style.STROKE
        wavePaint.strokeWidth = strokeWidth
        wavePaint.strokeCap = Paint.Cap.ROUND
        wavePaint.strokeJoin = Paint.Join.ROUND

        bgTrackPaint.color = Color.parseColor("#222838")
        bgTrackPaint.style = Paint.Style.STROKE
        bgTrackPaint.strokeWidth = strokeWidth
        bgTrackPaint.strokeCap = Paint.Cap.ROUND

        thumbPaint.color = Color.WHITE
        thumbPaint.style = Paint.Style.FILL

        thumbGlowPaint.color = Color.parseColor("#06B6D4")
        thumbGlowPaint.style = Paint.Style.FILL

        progressDrawable = ColorDrawable(Color.TRANSPARENT)
        thumb = ColorDrawable(Color.TRANSPARENT)
        splitTrack = false
        setBackgroundColor(Color.TRANSPARENT)
        currentAmplitude = 0f
    }

    fun setAccentColor(accentColor: Int) {
        wavePaint.color = accentColor
        thumbGlowPaint.color = accentColor
        invalidate()
    }

    override fun setOnSeekBarChangeListener(l: OnSeekBarChangeListener?) {
        super.setOnSeekBarChangeListener(l)
        this.listener = l
    }

    fun setPlaying(playing: Boolean) {
        if (this.isPlaying == playing) return
        this.isPlaying = playing
        animateStateTransition()
    }

    private fun animateStateTransition() {
        amplitudeAnimator?.cancel()

        val targetAmp = if (isPlaying) waveAmplitude else 0f
        amplitudeAnimator = ValueAnimator.ofFloat(currentAmplitude, targetAmp).apply {
            duration = 350
            addUpdateListener { animation ->
                currentAmplitude = animation.animatedValue as Float
                invalidate()
            }
            start()
        }

        if (isPlaying) {
            if (waveAnimator == null) {
                waveAnimator = ValueAnimator.ofFloat(0f, (2 * PI).toFloat()).apply {
                    duration = 1200
                    interpolator = LinearInterpolator()
                    repeatCount = ValueAnimator.INFINITE
                    addUpdateListener { animation ->
                        phase = animation.animatedValue as Float
                        invalidate()
                    }
                }
            }
            if (waveAnimator?.isRunning != true) {
                waveAnimator?.start()
            }
        } else {
            if (currentAmplitude <= 0.05f) {
                waveAnimator?.cancel()
            }
        }
    }

    @Synchronized
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val desiredHeight = (36f * density).roundToInt()
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        val heightSize = MeasureSpec.getSize(heightMeasureSpec)

        val measuredHeight = when (heightMode) {
            MeasureSpec.EXACTLY -> heightSize
            MeasureSpec.AT_MOST -> min(desiredHeight, heightSize)
            else -> desiredHeight
        }

        setMeasuredDimension(MeasureSpec.getSize(widthMeasureSpec), measuredHeight)
    }

    @Synchronized
    override fun onDraw(canvas: Canvas) {
        val w = width
        val h = height
        if (w <= 0 || h <= 0) return

        val centerY = h / 2f
        val trackStart = paddingLeft + thumbGlowRadius
        val trackEnd = w - paddingRight - thumbGlowRadius
        val trackWidth = max(1f, trackEnd - trackStart)

        val maxVal = max
        val progressFraction = if (maxVal > 0) max(0f, min(1f, progress.toFloat() / maxVal)) else 0f
        val thumbX = trackStart + trackWidth * progressFraction

        if (thumbX < trackEnd) {
            canvas.drawLine(thumbX, centerY, trackEnd, centerY, bgTrackPaint)
        }

        if (thumbX > trackStart) {
            val playedWidth = thumbX - trackStart
            if (currentAmplitude > 0.1f && playedWidth > 4f * density) {
                wavePath.reset()
                wavePath.moveTo(trackStart, centerY)

                val step = 2.0f * density
                val rampDist = min(waveLength * 0.75f, playedWidth / 2f)

                var x = trackStart
                while (x <= thumbX) {
                    val distStart = x - trackStart
                    val distEnd = thumbX - x
                    var env = 1f

                    if (rampDist > 0.1f) {
                        val startRamp = min(1f, distStart / rampDist)
                        val endRamp = min(1f, distEnd / rampDist)
                        val minRamp = min(startRamp, endRamp)
                        env = sin(minRamp * PI / 2.0).toFloat()
                    }

                    val angle = (2.0 * PI * (x - trackStart) / waveLength - phase).toFloat()
                    val y = centerY + currentAmplitude * env * sin(angle.toDouble()).toFloat()
                    wavePath.lineTo(x, y)
                    x += step
                }

                wavePath.lineTo(thumbX, centerY)
                canvas.drawPath(wavePath, wavePaint)
            } else {
                canvas.drawLine(trackStart, centerY, thumbX, centerY, wavePaint)
            }
        }

        val currentGlow = if (isDragging) (thumbGlowRadius + 3.0f * density) else thumbGlowRadius
        val glowAlpha = if (isDragging) 180 else 120
        thumbGlowPaint.alpha = glowAlpha
        canvas.drawCircle(thumbX, centerY, currentGlow, thumbGlowPaint)
        canvas.drawCircle(thumbX, centerY, thumbRadius, thumbPaint)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!isEnabled) return false

        val trackStart = paddingLeft + thumbGlowRadius
        val trackEnd = width - paddingRight - thumbGlowRadius
        val trackWidth = max(1f, trackEnd - trackStart)

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                isDragging = true
                parent?.requestDisallowInterceptTouchEvent(true)
                updateTouchProgress(event.x, trackStart, trackWidth)
                listener?.onStartTrackingTouch(this)
                invalidate()
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                if (isDragging) {
                    updateTouchProgress(event.x, trackStart, trackWidth)
                    invalidate()
                    return true
                }
            }
            MotionEvent.ACTION_UP -> {
                if (isDragging) {
                    updateTouchProgress(event.x, trackStart, trackWidth)
                    isDragging = false
                    listener?.onStopTrackingTouch(this)
                    performClick()
                    invalidate()
                    return true
                }
            }
            MotionEvent.ACTION_CANCEL -> {
                if (isDragging) {
                    updateTouchProgress(event.x, trackStart, trackWidth)
                    isDragging = false
                    listener?.onStopTrackingTouch(this)
                    invalidate()
                    return true
                }
            }
        }

        return super.onTouchEvent(event)
    }

    override fun performClick(): Boolean {
        super.performClick()
        return true
    }

    private fun updateTouchProgress(x: Float, trackStart: Float, trackWidth: Float) {
        val clampedX = max(trackStart, min(trackStart + trackWidth, x))
        val fraction = (clampedX - trackStart) / trackWidth
        val newProgress = (fraction * max).roundToInt()
        progress = newProgress
        listener?.onProgressChanged(this, newProgress, true)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        waveAnimator?.cancel()
        amplitudeAnimator?.cancel()
    }
}
